frontend/src/metabase/dashboard/components/DashCard/DashCard.jsx

```diff
@@ -42,6 +47,39 @@ const WrappedVisualization = WithVizSettingsData(
   connect(null, dispatch => ({ dispatch }))(Visualization),
 );
 
+function preventDragging(event) {
+  event.stopPropagation();
+}
+
+function getSeriesError(series) {
+  const isAccessRestricted = series.some(
+    s =>
+      s.error_type === SERVER_ERROR_TYPES.missingPermissions ||
+      s.error?.status === 403,
+  );
+
+  if (isAccessRestricted) {
+    return {
+      message: ERROR_MESSAGE_PERMISSION,
+      icon: "key",
+    };
+  }
+
+  const errors = series.map(s => s.error).filter(Boolean);
+  if (errors.length > 0) {
+    if (IS_EMBED_PREVIEW) {
+      const message = errors[0]?.data || ERROR_MESSAGE_GENERIC;
+      return { message, icon: "warning" };
+    }
+    return {
+      message: ERROR_MESSAGE_GENERIC,
+      icon: "warning",
+    };
+  }
+
+  return;
+}
```

Comment:
this feels like it should go in a `utils` or `lib/` file somewhere?

Reply:
I'm not sure if there's any other place in the app that'd actually use it, so I'd vote for keeping it as close to its consumer as possible